File: exercism/cpp/interest_is_interesting/interest_is_interesting.cpp
#include <iostream>


double interest_rate(double balance) {
    if (balance < 0)
        return 3.213;
    else if (balance >= 0 && balance < 1000)
        return 0.5;
    else if (balance >= 1000 and balance < 5000)
        return 1.621;
    else
        return 2.475;
}

double yearly_interest(double balance) {
    return balance * interest_rate(balance) / 100.0;
}

double annual_balance_update(double balance) {
    return balance + yearly_interest(balance);
}

int years_until_desired_balance(double balance, double target_balance) {
    int counter{0};

    while (balance < target_balance) {
        balance = annual_balance_update(balance);
        ++counter;
    }

    return counter;
}

File: exercism/csharp/InterestIsInteresting.cs
static class SavingsAccount
{
    public static float InterestRate(decimal balance)
    {
        return balance switch {
            < 0 => 3.213f,
            < 1000 => 0.5f,
            < 5000 => 1.621f,
            _ => 2.475f
        };
    }

    public static decimal Interest(decimal balance)
    {
        return balance * (decimal)InterestRate(balance) / 100;
    }

    public static decimal AnnualBalanceUpdate(decimal balance)
    {
        return balance + Interest(balance);
    }

    public static int YearsBeforeDesiredBalance(decimal balance, decimal targetBalance)
    {
        int n = 0;
        decimal currentBalance = balance;

        while (currentBalance < targetBalance)
        {
            currentBalance = AnnualBalanceUpdate(currentBalance);
            n++;
        }

        return n;
    }
}
